### ISSUE TITLE
[Feature request] Lock app with password and to prevent the app from being uninstalled and lock it with a password

### ISSUE THREAD

**Issue by elyoas** [IngoZenz/personaldnsfilter#263]

**Comment by lynxthecat** [comment]
@elyoas I agree and think that it is highly desirable to have a passphrase requested for any form of attempt to circumvent the DNS filter such as switching off the 'always on VPN'. But this is subject to what is allowed in a non-rooted android device, right? Unfortunately I have a feeling the latter may effectively prevent introducing a passphrase requirement to circumvent the DNS filter. And rooting an android devices seems like a hassle and presumably breaks important banking apps, etc. (or requires endless tweaking to keep them working).

**Comment by elyoas** [comment]
The app I mentioned is already on the app store, and it can already do all that is required to prevent a user from stopping it or uninstalling it.

This means it's possible even if the phone is non-rooted.

Perhaps it's a solution to use a separate app to do this kind of locking? But I haven't heard of any open source apps which provide these capabilities.. Thanks for your response.

**Comment by IngoZenz** [comment]
you can try https://play.google.com/store/search?q=norton+app+lock&c=apps

**Comment by lynxthecat** [comment]
But this isn't just about locking an app. It's about preventing a user from stopping or bypassing a VPN in Android, which seems like a pretty core feature. I can't imagine an app controlling that behaviour.

**Comment by elyoas** [comment]
I haven't used it for some time now so I don't remember the exact proceduce. But it locks the VPN settings with a PIN. Every time you open the settings menu which turns off or changes the VPN, it directs you to the app which requests a PIN.

The same happens every time you go to the apps (in settings), once you open that app (where it gives you the uninstall option), it redirects you to the app which requests a PIN.

The app is simple to use you can try it yourself, it is very effective and there is no way round it.

**Comment by Averell7** [comment]
I am also interested by this feature for the same reason (using personalDNSfilter for kids). I have tested in the past Qustodio which can be uninstalled only with a code. But the situation is probably different : I do not remember if they use a VPN. We may collaborate to the programming, but this needs some ideas of where to start.

**Comment by IngoZenz** [comment]
Currently pDNSf only supports the password lock for displaying and changing app settings, but how to prevent it to be uninstalled or VPN stopped, I have no idea.
